8251158: Implementation of JEP 387: Elastic Metaspace

tstuefe opened the pull request:
Hi all,

this is the continuation of the ongoing review for the JEP387 implementation (last rounds see [1] [2]). Sorry for the delay, had vacation then the entrance of Skara delayed things a bit.

For the delta diff please see [3].

This is the first time I do a large PR after Skara, so if something is wrong please bear with me. I cannot answer all feedback individually in this PR body, but I incorporated almost all into the new revision.

What changed since the last version:

- I renamed most metaspace files back to the original naming scheme or to something similar, hopefully capturing the group consent.

- I changed the way allocation guards are checked if MetaspaceGuardAllocations is enabled. Before, I would test for overwrites upon CLD destruction, but since that check was subject to VerifyMetaspaceInterval it only ran for every nth class loader which made it rather pointless. Now I run it always.

- I also improved the printout on block corruption, and log block corruption unconditionally before asserting.

- I also fixed up and commented the death test which tests for allocation overwriters (test_allocationGuard.cpp)

Side note, I find the corruption check very useful but if you guys think it is too much I still can remove the feature.

- In ChunkManager::purge() I improved the comments after discussions with Leo.

- I fixed a bug with VerifyMetaspaceInterval: if set to 1 the "SOMETIMES" sections were supposed to fire always, but due to a one-off error they only fired every second time. Now, if -XX:VerifyMetaspaceInterval=1, the checks really run every time.

- Fixed indentation issues as Leo requested

- Rewrote the condition and the assert in VirtualSpaceList::allocate_root_chunk() as Leo requested

- I removed the "can_purge" logic from VirtualSpaceList. The list does not need to know. It just should iterate all nodes and attempt purging, and if a node does not own its ReservedSpace, it refuses to be purged. That is simpler and more flexible since it allows us to have list with purge-able and non-purge-able nodes.

- and various smaller fixes, mainly on request of Leo.

@lkorinth:

> VirtualSpaceNode.hpp
>
>102   // Start pointer of the area.
>103   MetaWord* const _base;
>
>How does this differ from _rs._base? Really needed?
>
>105   // Size, in words, of the whole node
>106   const size_t _word_size;
>
>Can we not calculate this from _rs.size()?

You are right, _base and _word_size are directly related to the underlying space. But I'd prefer to leave it the way it is. Mainly because ReservedSpace::_base and ::_size are nonconst and theoretically can change under me. It is highly improbable but I'd like to know. Note that VirtualSpaceNode::verify checks that.

Should we clean up ReservedSpace at some point and make those members const - as they should be - then I would rewrite this as you suggest.

Thanks, again, for all your review work!

------


[1] https://mail.openjdk.java.net/pipermail/hotspot-runtime-dev/2020-August/041162.html
[2] https://mail.openjdk.java.net/pipermail/hotspot-runtime-dev/2020-September/041628.html
[3] https://github.com/openjdk/jdk/commit/731f795bc0c1c502dc6cac8f866ff45a15bdd02d


/issue 8251158
/cc hotspot-runtime, hotspot-gc
/test tier1, tier2, tier3, tier4, tier5, tier6, tier7, tier8

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8251158](https://bugs.openjdk.java.net/browse/JDK-8251158): Implementation of JEP 387: Elastic Metaspace


### Reviewers
 * [Leo Korinth](https://openjdk.java.net/census#lkorinth) (@lkorinth - Committer) ⚠️ Review applies to ece884f1c6f51c93331a68e31156880c2661e3db
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to b6e8b842cb737e087c59a8ef3fb70a708262d62c
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to f5cf615bbb0154c915dcea8bb9574f88d763ce69
 * [Richard Reingruber](https://openjdk.java.net/census#rrich) (@reinrich - Committer) ⚠️ Review applies to e895bfa0ee7a66f6736a36fcead79fd4e106d3ff


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/336/head:pull/336`
`$ git checkout pull/336`

tstuefe commented:
/test

stefank commented on src/hotspot/share/memory/metaspace/virtualSpaceNode.cpp:
```diff
@@ -25,564 +26,503 @@
 #include "runtime/os.hpp"
-#include "services/memTracker.hpp"
-#include "utilities/copy.hpp"
+
```
Just a drive-by comment. Please don't introduce this new style of adding newlines between our internal include lines. The same comment about the newlines after precompiled.hpp. I know that the Shenandoah code already does that, and I've seen some of the metaspace code do that as well, but please stick to the set standard way of listing our includes. This applies to all files in this patch.

tstuefe replied:
Sure, no problem. I changed it.

stefank replied:
Much obliged

lkorinth reviewed (changes requested):
I just have a few cosmetic comments. Otherwise it looks good to me.

If you sort by ASCII order, I get a few unordered includes:
("." < "/" < "a-z,A-Z")


In file b/src/hotspot/share/gc/shared/genCollectedHeap.cpp:
[not sorted]:
#include "memory/metaspaceCounters.hpp" (Context)
#include "memory/metaspace/metaspaceSizesSnapshot.hpp" (Addition)
#include "memory/resourceArea.hpp" (Context)

In file b/src/hotspot/share/memory/metaspace.cpp:
[not sorted]:
#include "memory/metaspace/virtualSpaceList.hpp" (Context)
#include "memory/metaspace.hpp" (Addition)
#include "memory/metaspaceShared.hpp" (Context)

In file b/src/hotspot/share/memory/metaspace/chunkManager.cpp:
[not sorted]:
#include "memory/metaspace/metaspaceCommon.hpp" (Context)
#include "memory/metaspace/metaspaceContext.hpp" (Addition)
#include "memory/metaspace/metachunk.hpp" (Addition)

In file b/src/hotspot/share/memory/metaspace/chunkManager.cpp:
[not sorted]:
#include "memory/metaspace/metaspaceContext.hpp" (Addition)
#include "memory/metaspace/metachunk.hpp" (Addition)
#include "memory/metaspace/metaspaceSettings.hpp" (Addition)


blockTree.hpp
add a space after loop keyword "for(;;) {" -> "for (;;) {"

blockTree.cpp
add a space after loop keyword "} while(0)" -> "} while (0)" (twice in file)

I still think we should try to get the initializer list indented
somewhat consistently. (This is really boring, and hard as we do not
have precise indentation rules and no mechanical indenter). Sorry for
mentioning this, but now might be the time to get indentation
consistent at least in metaspace. The indentation level seems to most
often be 2 or 4. Sometimes "Haskell" indentation with "," at the
beginning of each line is used (I like this, but I think I am in a
small minority) and most often it is not used. Sometimes several field
members are initialised on multi line initializer lists, sometimes
only one per line (I prefer one per line, if not all fits on a single
line). Open curly braces seem to mostly (but not always) be on a new line.

If you could normalize the style I would be happy, I yield my own
preferences to you and the other reviewers on what style to use, so
that we have a possibility to agree :-)

Thanks, Leo

lkorinth commented:
Regarding the order of "." and "/" take what you prefer.

coleenp reviewed (approved):
Besides my comment about globals.hpp option MetaspaceGuardAllocations, my comments are minor things and I approve this change.  There might be some additional things we find that we'll want to change once this code is integrated.  This is a significant improvement to metaspace memory management.  Great work, @tstuefe !

tstuefe commented:
Hi Leo,

> I just have a few cosmetic comments. Otherwise it looks good to me.
> 
> If you sort by ASCII order, I get a few unordered includes:
> ("." < "/" < "a-z,A-Z")
> ....

I will fix all those.

> 
> blockTree.hpp
> add a space after loop keyword "for(;;) {" -> "for (;;) {"
> 
> blockTree.cpp
> add a space after loop keyword "} while(0)" -> "} while (0)" (twice in file)
> 

Okay

> I still think we should try to get the initializer list indented
> somewhat consistently.

I agree, will do.

tstuefe commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -0,0 +1,207 @@
+  size_t lim2; // )
+};
+
+void BlockTree::verify() const {
```
> I think this sort of deep verification should be in a gtest instead, at least for blockTree. Note, this can be fixed/discussed in a future RFE.

:( This would affect a lot of my code since I do a lot of in-file verifications. Yeah, lets discuss this separately.

tstuefe commented on src/hotspot/share/memory/metaspace/classLoaderMetaspace.cpp:
```diff
@@ -0,0 +1,202 @@
+  : _lock(lock)
+  , _space_type(space_type)
+  , _non_class_space_arena(NULL)
+  , _class_space_arena(NULL)
```
Yes, I agree with you two. I will find a common form and fix all my initializations.

tstuefe commented on src/hotspot/share/memory/metaspace/commitMask.hpp:
```diff
@@ -0,0 +1,178 @@
+           word_size, _words_per_bit);
+    check_pointer(start + word_size - 1);
+  }
+#endif
```
Ok

tstuefe commented on src/hotspot/share/memory/metaspace/metachunk.hpp:
```diff
@@ -21,152 +22,292 @@
+  void initialize(VirtualSpaceNode* node, MetaWord* base, chunklevel_t lvl) {
+    _vsnode = node; _base = base; _level = lvl;
+    _used_words = _committed_words = 0; _state = State::Free;
+    _next = _prev = _next_in_vs = _prev_in_vs = NULL;
```
Yes, I'll reuse clear() instead.

tstuefe commented on src/hotspot/share/memory/metaspace/metachunkList.hpp:
```diff
@@ -0,0 +1,102 @@
+
+  void add(Metachunk* c) {
+    // Note: contains is expensive (linear search).
+    ASSERT_SOMETIMES(contains(c) == false, "Chunk already in this list");
```
Ok

tstuefe commented on src/hotspot/share/memory/metaspace/metachunkList.hpp:
```diff
@@ -0,0 +1,102 @@
+#ifndef SHARE_MEMORY_METASPACE_METACHUNKLIST_HPP
+#define SHARE_MEMORY_METASPACE_METACHUNKLIST_HPP
+
+#include "memory/metaspace/counters.hpp"
```
Yes, since we still use IntCounter.

tstuefe commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -1580,6 +1579,18 @@ const intx ObjectAlignmentInBytes = 8;
+          "options: balanced, aggressive, none")                            \
+                                                                            \
+  product(bool, MetaspaceGuardAllocations, false,                           \
+          "Metapace allocations are guarded.")                              \
```
Ok!

tstuefe commented:
Hi,

thanks for the reviews! New version:

- Hotspot style fixes - sorted all include sections, and removed empty lines from them; fixed whitespace issues in loop constructs; fixed include guard names; squashed empty lines
- Moved verification code from commitmask.hpp to commitmask.cpp
- Squashed multiple empty lines to one and condensed code a bit more
- Changed ctor initializer lists: I decided on two formats. Two space indentation, colon follows prototype name, punctuation at the end, all in one line or one line per member:

XXX() : _m1(xx), _m2(xx) {
  blub
}

XXX() :
  _m1(xx),
  _m2(xx),
  _m3(xx),
  _m4(xx)
{
  blub
}

- Made MetaspaceGuardAllocations diagnostic

tstuefe commented:
> Besides my comment about globals.hpp option MetaspaceGuardAllocations, my comments are minor things and I approve this change. There might be some additional things we find that we'll want to change once this code is integrated. This is a significant improvement to metaspace memory management. Great work, @tstuefe !

Thank you Coleen! :)

iklam commented on src/hotspot/share/memory/metaspace.hpp:
```diff
@@ -234,212 +150,115 @@ class Metaspace : public AllStatic {
+//  non-class space).
+//
+// ClassLoaderMetaspace only exists to hide this logic from upper layers:
+//
```
I would suggest rewriting the comments to

```
// A ClassLoaderMetaspace manages  MetaspaceArena(s) for a CLD.
//
// A CLD owns one MetaspaceArena if UseCompressedClassPointers is false. Otherwise
// it owns two -- one for the Klass* objects from the class space, one for the other types
// of MetaspaceObjs from the non-class space.
```

(I think "hide this logic ...." can be omitted. We have lots of abstractions so there's no need to explicitly call it out).

tstuefe replied:
Okay, sounds good.

iklam reviewed (changes requested):
So far I have only reviewed the C code that uses the metaspace, but not the metaspace implementation itself (memory/metaspace*, memory/metaspace/*). I also reviewed a bit of classes.

tstuefe commented on test/hotspot/jtreg/runtime/cds/MaxMetaspaceSize.java:
```diff
@@ -44,9 +44,8 @@ public static void main(String[] args) throws Exception {
 
     if (Platform.is64bit()) {
-      processArgs.add("-XX:MaxMetaspaceSize=3m");
+      processArgs.add("-XX:MaxMetaspaceSize=8m");
```
On the contrary. Minimum metaspace usage to run a simple HelloWorld went way down. 

Before: 10,00 MB reserved,       4,75 MB ( 48%) committed 
Now: 12,00 MB reserved,     448,00 KB (  3%) committed.

mainly because now we commit the large chunk for the boot loader only on demand. Reservation (vsize) numbers are somwhat chunkier now since the default size for a VirtualSpaceNode went up from 4 to 8MB.

You now can get away with -XX:MaxMetaspaceSize=~500K to start a simple program.

While writing this, I wondered why I made this change to your test. It is not needed for the current version, probably a remnant from some earlier prototype. I will revert this line and check the other tests too.

tstuefe commented on src/hotspot/share/gc/parallel/psParallelCompact.cpp:
```diff
@@ -1058,7 +1058,7 @@ void PSParallelCompact::post_compact()
   // Delete metaspaces for unloaded class loaders and clean up loader_data graph
   ClassLoaderDataGraph::purge(/*at_safepoint*/true);
-  MetaspaceUtils::verify_metrics();
+  DEBUG_ONLY(MetaspaceUtils::verify();)
```
Unless you really want me to, or other Reviewers chime in, I'd rather leave it as it is. I prefer this style since it is clear at the callsites that this is only ASSERT code.

tstuefe commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -1580,6 +1579,18 @@ const intx ObjectAlignmentInBytes = 8;
                                                                             \
+  product(uintx, ForceCompressedClassSpaceStartAddress, 0, EXPERIMENTAL,    \
+          "Force class space start address to a given value.")              \
+                                                                            \
```
Good catch, will remove.

tstuefe commented on src/hotspot/share/jvmci/jvmciCompilerToVM.hpp:
```diff
@@ -25,6 +25,7 @@
 #define SHARE_JVMCI_JVMCICOMPILERTOVM_HPP
 
 #include "gc/shared/cardTable.hpp"
+#include "gc/shared/collectedHeap.hpp"
```
Sure.

tstuefe commented on test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/LargePages.java:
```diff
@@ -46,8 +46,8 @@ public static void test(String[] args) throws Exception {
         SharedStringsUtils.dump(TestCommon.list("HelloString"),
             "SharedStringsBasic.txt", CDS_LOGGING,
-            "-XX:+UseLargePages", "-XX:+UseLargePagesInMetaspace");
+            "-XX:+UseLargePages");
```
I missed that. Thanks.

iklam commented on src/hotspot/share/memory/metaspace.hpp:
```diff
@@ -234,212 +150,113 @@ class Metaspace : public AllStatic {
-  // class loader and hidden class loader.
-  void initialize_first_chunk(Metaspace::MetaspaceType type, Metaspace::MetadataType mdtype);
-  metaspace::Metachunk* get_initialization_chunk(Metaspace::MetaspaceType type, Metaspace::MetadataType mdtype);
+  // A reference to an outside lock, held by the CLD.
```
Should this be moved to classLoaderMetaspace.hpp (now that you have created a new file classLoaderMetaspace.cpp).

iklam commented on src/hotspot/share/memory/metaspace/metaspaceSettings.hpp:
```diff
@@ -0,0 +1,90 @@
+  DEBUG_ONLY(static bool _use_allocation_guard;)
+
+  // If true, we handle deallocated blocks (default).
+  DEBUG_ONLY(static bool _handle_deallocations;)
```
I think "handle deallocated blocks" is a bit unclear. How about "If false, no blocks are ever freed and left-over space in chunks are not salvaged.".

iklam commented on src/hotspot/share/memory/metaspace/metaspaceArena.cpp:
```diff
@@ -0,0 +1,542 @@
+            "Requested size too large (" SIZE_FORMAT ") - max allowed size per allocation is " SIZE_FORMAT ".",
+            requested_word_size, chunklevel::MAX_CHUNK_WORD_SIZE);
+
+  const int growth_step = _chunks.count();
```
growth_step is unused.

iklam reviewed (approved):
LGTM

tstuefe commented:
> LGTM

Thanks Ioi!

I moved CLMS to an own header as you suggested. I decided to put them outside the metaspace namespace and the metaspace subdirectory since they belong to the external interface (strictly speaking they are the glue between CLD and metaspace). Also keeps changes to outside code small.

Cheers, Thomas

lkorinth reviewed (approved):
Let us ship this. Thanks for all your great work, great presentations and for doing all the tedious fixes.

/Leo

tstuefe commented:
> Let us ship this. Thanks for all your great work, great presentations and for doing all the tedious fixes.
> 
> /Leo

Great, thank you too for your thorough review work!

reinrich commented:
Hi Thomas, I'm currently working on a review. There are still many sections where you separate 2 lines of code with an empty line. My review is filling up with comments on that. I'm afraid this will hide the few other/real points therefore I'll stop annotating empty lines. Instead I'd like to ask you to look over each file changed (you cannot see it always in the diff on Github) and remove as many empty lines as possible to better match the existing style. `virtualSpaceNode.cpp` would be a good file to start with (e.g. lines 106 and following, 280 and following). You could also grep for { followed by an empty line and } proceeded by an empty line. Thanks, Richard.

reinrich commented:
> 
> 
> _Mailing list message from [Thomas St��fe](mailto:thomas.stuefe@gmail.com) on [hotspot-gc-dev](mailto:hotspot-gc-dev@openjdk.java.net):_
> 
> > <snip>
> > If it is important to you I will remove the empty lines even though I
> > think it hurts readability. But I am eager to ship this PR, so please let
> > us concentrate on real issues. If you have still concerns let us discuss
> > those.
> 
> Okay, I removed some empty lines to move forward.
> 
> Thanks, Thomas
> 
> >

Thank's for doing it,
Richard.

reinrich reviewed (changes requested):
Hi Thomas,

I've viewed 41/166 files. I'm sending you my findings so far so we can work concurrently. Also I'm curious how it goes sending a partial review and then continuing.

Probably I won't manage to review all the tests but I do hope I get to read all files of the implementation. Working on it now...

Btw: I couldn't find out how to navigate from one of my comments to the line of code. At least, when I click on the file name, this brings me to a view of the patch at the version I commented on so the line numbers are correct even if they changed in the head revision of the pr.

Cheers, Richard.

tstuefe commented:
> Hi Thomas,
> 
> I've viewed 41/166 files. I'm sending you my findings so far so we can work concurrently. Also I'm curious how it goes sending a partial review and then continuing.

You just finish one Review (with "comment" or "reqzest changes"), and later start a new Review. Only downside is that you get multiple mails in the ML, but I think that is okay here.

> 
> Probably I won't manage to review all the tests but I do hope I get to read all files of the implementation. Working on it now...
> 

Totally fine. The implementation in hotspot is the important part.

> Btw: I couldn't find out how to navigate from one of my comments to the line of code. At least, when I click on the file name, this brings me to a view of the patch at the version I commented on so the line numbers are correct even if they changed in the head revision of the pr.

The line of code a comment applies to is the lowest in the code excerpt shown atop of the comment. Unless the comment applied to multiple lines, then it shows (eg "Comment on lines +105 to +110").

> 
> Cheers, Richard.

Thanks, Tomas

tstuefe commented on src/hotspot/share/services/virtualMemoryTracker.cpp:
```diff
@@ -671,10 +671,9 @@ void MetaspaceSnapshot::snapshot(Metaspace::MetadataType type, MetaspaceSnapshot
+  mss._free_in_bytes[type] = mss._committed_in_bytes[type] - mss._used_in_bytes[type];
 }
 
 void MetaspaceSnapshot::snapshot(MetaspaceSnapshot& mss) {
```
Looks like it. I opened https://bugs.openjdk.java.net/browse/JDK-8254012.

tstuefe commented on src/hotspot/share/memory/classLoaderMetaspace.cpp:
```diff
@@ -0,0 +1,203 @@
+        "class sm");
+  }
+
+  UL2(debug, "born (SpcMgr nonclass: " PTR_FORMAT ", SpcMgr class: " PTR_FORMAT ".",
```
Right.

tstuefe commented on src/hotspot/share/memory/metaspace/binList.hpp:
```diff
@@ -0,0 +1,226 @@
+    assert(index >= 0 && index < num_lists, "Invalid index %d", index);
+    int i2 = index;
+    mask_t m = _mask >> i2;
+    if (m > 0) {
+      // count leading zeros would be helpful.
+      while ((m & 1) == 0) {
+        assert(_blocks[i2] == NULL, "mask mismatch");
+        i2++;
+        m >>= 1;
+      }
+      // We must have found something.
+      assert(i2 < num_lists, "sanity.");
+      assert(_blocks[i2] != NULL, "mask mismatch");
+      return i2;
+    }
```
Right. I'll just remove the comment.

reinrich replied:
It's not just the comment. I think you should replace the while-loop with [count_trailing_zeros()](https://github.com/openjdk/jdk/blob/d296708ca66ab15e31a05f963a5e162a3e3f07f9/src/hotspot/share/utilities/count_trailing_zeros.hpp#L31).

tstuefe commented on src/hotspot/share/memory/metaspace.hpp:
```diff
@@ -215,11 +130,10 @@ class Metaspace : public AllStatic {
 
   static void report_metadata_oome(ClassLoaderData* loader_data, size_t word_size,
-                                   MetaspaceObj::Type type, MetadataType mdtype, TRAPS);
+                                   MetaspaceObj::Type type, Metaspace::MetadataType mdtype, TRAPS);
```
The enums moved around a lot until ending up at the exact place as before the patch. I'll remove the scope.

tstuefe commented on src/hotspot/share/memory/metaspace/blockTree.hpp:
```diff
@@ -0,0 +1,410 @@
+        succ = succ->_left;
+      }
+    } else {
+      succ = n->_parent;
+      Node* n2 = n;
+      // As long as I am the right child of my parent, search upward
+      while (succ != NULL && n2 == succ->_right) {
+        n2 = succ;
+        succ = succ->_parent;
+      }
```
I rather leave it, I find it clearer that way.

tstuefe commented on src/hotspot/share/memory/metaspace/blockTree.hpp:
```diff
@@ -0,0 +1,410 @@
+  }
+
+  // Given a node n and an insertion point, insert n under insertion point.
+  void insert(Node* insertion_point, Node* n) {
```
True.

tstuefe commented on src/hotspot/share/memory/classLoaderMetaspace.cpp:
```diff
@@ -0,0 +1,203 @@
+
+// Prematurely returns a metaspace allocation to the _block_freelists
+// because it is not needed anymore.
+void ClassLoaderMetaspace::deallocate(MetaWord* ptr, size_t word_size, bool is_class) {
+
+  Metaspace::assert_not_frozen();
+
+  if (use_class_space(is_class)) {
+    class_space_arena()->deallocate(ptr, word_size);
+  } else {
+    non_class_space_arena()->deallocate(ptr, word_size);
+  }
+
+  DEBUG_ONLY(InternalStats::inc_num_deallocs();)
+
+}
```
Okay will fix.

tstuefe commented on src/hotspot/share/memory/metaspace/blockTree.hpp:
```diff
@@ -0,0 +1,410 @@
+    }
+  }
+
+  // Given a node n, return its predecessor in the tree
+  // (node with the next-smaller size).
+  static Node* predecessor(Node* n) {
+    Node* pred = NULL;
+    if (n->_left != NULL) {
+      pred = n->_left;
+      while (pred->_right != NULL) {
+        pred = pred->_right;
+      }
+    } else {
+      pred = n->_parent;
+      Node* n2 = n;
+      while (pred != NULL && n2 == pred->_left) {
+        n2 = pred;
+        pred = pred->_parent;
+      }
+    }
+    return pred;
+  }
+
```
Will remove.

tstuefe commented on src/hotspot/share/memory/metaspace/freeChunkList.hpp:
```diff
@@ -0,0 +1,272 @@
+// Should we ever encounter situations where the O(n) search is a bottleneck, this
+//  structure can easily be optimized (e.g. a BST). But for now lets keep this simple.
+
+class FreeChunkList {
```
Yes, this could possibly be done. Its a bit of a matter of taste. In this case note that FreeChunkList orders chunks by commit state on insert, and does other stuff particular to that list, so you will need to revive that functionality. As it is now I find this coding clear and easy to read. Any solution we arrive at after factoring out should not be worse.

reinrich replied:
I have not yet looked closely at freeChunkList.[ch]pp. Maybe `FreeChunkList` isn't a good candidate for doing it. I still think there is potential for reuse. I'm ok with postponing it though.

tstuefe commented on src/hotspot/share/memory/metaspace/chunkManager.hpp:
```diff
@@ -26,173 +27,162 @@
+  ChunkManager(const char* name, VirtualSpaceList* space_list);
+
+  // On success, returns a chunk of level of <preferred_level>, but at most <max_level>.
+  //  The first first <min_committed_words> of the chunk are guaranteed to be committed.
```
ok

tstuefe commented on src/hotspot/share/memory/metaspace/chunkManager.cpp:
```diff
@@ -21,622 +22,465 @@
+
+  DEBUG_ONLY(chunklevel::check_valid_level(max_level);)
+  DEBUG_ONLY(chunklevel::check_valid_level(preferred_level);)
+  assert(max_level >= preferred_level, "invalid level.");
```
True. 

(The line numbers are off in this snippet btw. The real code is at 125ff. Not sure what happened.)

reinrich replied:
> True.
> 
> (The line numbers are off in this snippet btw. The real code is at 125ff. Not sure what happened.)

This is what I meant in my review-message:

> > Btw: I couldn't find out how to navigate from one of my comments to the line of code. At least, when I click on the file name, this brings me to a view of the patch at the version I commented on so the line numbers are correct even if they changed in the head revision of the pr.
> 
> The line of code a comment applies to is the lowest in the code excerpt shown atop of the comment. Unless the comment applied to multiple lines, then it shows (eg "Comment on lines +105 to +110").
> 

If you click on the `chunkManager.cpp` in the comment header, this will bring you to the version of the file I commented on, and the line numbers there should match the numbers in the comment's code snippet. You will also see the this comment there.

tstuefe commented on src/hotspot/share/memory/metaspace/rootChunkArea.cpp:
```diff
@@ -0,0 +1,560 @@
+  // D D C B  <- first half split again ...
+  //
+
+  // As an optimization, since we usually do not split once but multiple times,
+  // to not do each split separately, since we would have to wire up prev_in_vs/next_in_vs
+  // on every level just to tear it open in the next level when we reintroduce a new
+  // half chunk splinter.
+  // Instead, just split split split and delay building up the double linked list of the
+  // new chunks at the end of all splits.
```
Right, good catch, this was for an earlier version. Found the original code too hard to read and too error prone. I'll remove the comment.

reinrich reviewed (changes requested):
Hi Thomas,

this is batch 2. I'm now at 61/166 files :)

Cheers, Richard.

tstuefe commented on src/hotspot/share/memory/metaspace/binList.hpp:
```diff
@@ -0,0 +1,225 @@
+    int i2 = index;
+    mask_t m = _mask >> i2;
+    if (m > 0) {
+      while ((m & 1) == 0) {
+        assert(_blocks[i2] == NULL, "mask mismatch");
+        i2++;
+        m >>= 1;
+      }
```
Good point. Lets do this in a follow-up though.

reinrich replied:
Honestly, without this the optimization is questionable. Iterating the bits in `_mask` is like iterating the list heads. `is_empty()` can be easily optimized with a counter, but even this is questionable. The change is tiny and easy to test: just leave the while-loop as an assertion and run some class redefinition tests then remove the loop.

I'd object postponing it.

tstuefe replied:
Okay. I had my doubt about this optimization too, as had Leo. I'll just scrap it altogether, this will make the code more readable too.

tstuefe commented on src/hotspot/share/memory/metaspace/metaspaceArena.cpp:
```diff
@@ -0,0 +1,493 @@
+  return _growth_policy->get_level_at_step(growth_step);
+}
+
+// Given a chunk, add its remaining free committed space to the free block list.
+void MetaspaceArena::salvage_chunk(Metachunk* c) {
+  if (Settings::handle_deallocations() == false) {
+    return;
+  }
+
+  assert_lock_strong(lock());
+  size_t remaining_words = c->free_below_committed_words();
+  if (remaining_words > FreeBlocks::MinWordSize) {
+
+    UL2(trace, "salvaging chunk " METACHUNK_FULL_FORMAT ".", METACHUNK_FULL_FORMAT_ARGS(c));
+
+    MetaWord* ptr = c->allocate(remaining_words);
+    assert(ptr != NULL, "Should have worked");
+    _total_used_words_counter->increment_by(remaining_words);
+
+    add_allocation_to_fbl(ptr, remaining_words);
+
+    // After this operation: the chunk should have no free committed space left.
+    assert(c->free_below_committed_words() == 0,
+           "Salvaging chunk failed (chunk " METACHUNK_FULL_FORMAT ").",
+           METACHUNK_FULL_FORMAT_ARGS(c));
+  }
+}
```
Yes, but in practice, the benefit would be slim:
- it is very rare that you retire a chunk which is not even halfway used. For that to happen a metaspace allocation must be requested larger than half of the current chunk. Does not happen that often.
- This only affects the committed portion of the chunk. So, for larger chunks - where this could hurt - the waste (waste in the sense as "unnecessarily earmarked for one loader only") mostly consists of address space only, not committed space.

tstuefe replied:
Okay, I did some statistics and salvaging chunks which are not at least half-way used is super rare. I opt in this case for just leaving the code as it is since splitting the salvaged chunk would mean complicating the code, apart from drawing the central metaspace context lock.

reinrich replied:
Ok. Thanks for doing it.

tstuefe commented on src/hotspot/share/memory/metaspace/counters.hpp:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  void decrement() {
+#ifdef ASSERT
+    T old = Atomic::load_acquire(&_c);
+    assert(old >= 1,
+        "underflow (" UINT64_FORMAT "-1)", (uint64_t)old);
+#endif
```
Okay, I feel stupid now but I do not see it.

I see the non-atomicity, but not how it could cause a false positive assert here. I see that I could miss an assert, but not the other way around.

tstuefe replied:
I will remove the asserts. These classes are scheduled to become general utility classes anyway in a followup RFE, so they will be revisited soon.

reinrich replied:
It is not as clear as I though. So no need at all to feel stupid.

> I will remove the asserts. These classes are scheduled to become general utility classes anyway in a followup RFE, so they will be revisited soon.

I'm ok with that.

For the sake of completeness: I'd still think false positives are possible. Please look at the following example:

````
   1	_initialized = 0;
   2	_c = 0;
   3	
   4	Thread A                                  Thread B
   5	
   6	                                          Atomic::inc(&_c)
   7	
   8	                                          <membar StoreLoad|StoreStore>
   9	
  10	                                          _initialized = 1;
  11	
  12	while(_initialized == 0) busy_wait();
  13	
  14	T old = Atomic::load_acquire(&_c);
  15	
  16	assert(old >= 1, "underflow (" UINT64_FORMAT "-1)", (uint64_t)old);
  17	
  18	<fence>
  19	
  20	Atomic::dec(&_c);
````

Loading `old` at line 14 can be done speculatively before the loop, for instance at line 5. There is no barrier that would prevent this. `Atomic::inc` and `Atomic::dec` both have a `<fence>` before and a `<membar StoreLoad|StoreStore>` after the operation. Because of the effect of the `<fence>` in line 18 the `dec()` in line 20 cannot be done before the while in line 12. An underflow is not possible. The assertion can fail though.

So probably you could fix the assertion by adding a fence before the load of the `old` value.

tstuefe commented on src/hotspot/share/memory/metaspace/metachunk.cpp:
```diff
@@ -23,149 +24,285 @@
 
-size_t Metachunk::free_word_size() const {
-  return pointer_delta(end(), _top, sizeof(MetaWord));
+// Uncommit chunk area. The area must be a common multiple of the
+// commit granule size (in other words, we cannot uncommit chunks smaller than
+// a commit granule size).
+void Metachunk::uncommit() {
+  MutexLocker cl(MetaspaceExpand_lock, Mutex::_no_safepoint_check_flag);
+  uncommit_locked();
```
It is used in a number of gtests. I'd prefer to keep it since otherwise I would have to lock manually each time there, and also because of symmetry with other functions.

reinrich replied:
Ok, I see.

tstuefe commented on src/hotspot/share/memory/metaspace/freeChunkList.hpp:
```diff
@@ -0,0 +1,272 @@
+//    commit granule, so 99% of all searches would end at the first chunk (which is either
+//    uncommitted or committed to at least one commit granule size).
+//  - in all likelihood chunks, when added to this list, are either fully committed
+//    or fully uncommitted (see Settings::uncommit_on_return_min_word_size()).
```
Leftover. Will remove the reference.

tstuefe commented on src/hotspot/share/memory/metaspace/freeChunkList.hpp:
```diff
@@ -0,0 +1,272 @@
+// Beyond adding at either front or at back, we do not sort on insert, since the
+//  insert path is used during Metaspace reclamation which may happen at GC pause.
+//
+// During retrieval (at class loading), we search the list for a chunk of at least
+//  n committed words to satisfy the caller requested committed word size. We stop
+//  searching at the first fully uncommitted chunk.
+//
```
No, this is still valid, see FreeChunkListVector::search_chunk_ascending/desdending.

tstuefe replied:
No wait, the last sentence is outdated. Will fix.

reinrich replied:
> 
> 
> No wait, the last sentence is outdated. Will fix.

It is not just the last sentence.

> 
> 
> No, this is still valid, see FreeChunkListVector::search_chunk_ascending/desdending.

These methods look at the first chunk of every list. No list is searched.

But it's good. I don't want to be too badly pedantic.

tstuefe replied:
Fixed with https://github.com/openjdk/jdk/pull/336/commits/2c97008fcc526989f49002e88abeea3d1f3dcdd5

tstuefe commented on src/hotspot/share/memory/metaspace/metachunk.hpp:
```diff
@@ -21,152 +22,289 @@
+  // This can happen if a neighboring chunk is added or removed.
+  // This means only read or modify these members under expand lock protection.
+  Metachunk* _prev_in_vs;
+  Metachunk* _next_in_vs;
```
The idea is good, but at the moment _next_in_vs==NULL marks the last chunk within one root chunk area, and is used e.g. as loop stop. I would have to make Metachunk aware of the root chunk it lives in, which I rather avoid.

reinrich replied:
Ok.

tstuefe replied:
Thought about this a bit more, your proposal does not work.

Metachunk are stored in C-Heap and are physically separated from the chunks they represent. _next_in_vs points to the MetaChunk header **representing** the chunk following the current chunk. Sort of like this:

```
+---------+                       +---------+                       +---------+
|MetaChunk| <--next/prev_in_vs--> |MetaChunk| <--next/prev_in_vs--> |MetaChunk|
+---------+                       +---------+                       +---------+ 
    |                                 |                                  |
   base                              base                               base
    |                                 /                                  | 
   /        --------------------------                                  /
  /        /          -------------------------------------------------- 
 |         |         /
 v         v         v  
+---------+---------+---------+---------+
|  chunk  |  chunk  |  chunk  |  chunk  |
+---------+---------+---------+---------+
```

Given a pointer to the start of a chunk (or, payload), there is no way to get to the MetaChunk representing it. Therefore Metachunk::end() is no help. Therefore I need to store _prev_in_vs and _next_in_vs in the MetaChunk Header.

tstuefe commented on src/hotspot/share/memory/metaspace/metachunk.hpp:
```diff
@@ -21,152 +22,289 @@
+  bool ensure_committed(size_t new_committed_words);
+  bool ensure_committed_locked(size_t new_committed_words);
+
+  bool ensure_fully_committed()           { return ensure_committed(word_size()); }
+  bool ensure_fully_committed_locked()    { return ensure_committed_locked(word_size()); }
```
Okay will remove.

reinrich reviewed (changes requested):
Hi Thomas,

this is batch 3 with another 3 points. I'm sending it now because I think I cannot reply to your comments otherwise.

Cheers, Richard.

coleenp commented on src/hotspot/share/memory/metaspace/metachunk.hpp:
```diff
@@ -100,9 +125,46 @@ class VirtualSpaceNode;
+// +---------+---------+-------------------+
+// |  chunk  |  chunk  |      chunk        |
+// +---------+---------+-------------------+
+//
```
Do you have this same picture twice?  Other than that, thank you for the ascii art.  I think it'll help understand all of this code.

tstuefe replied:
Thought you like it :) The first picture describes the general division between headers and (possibly distant) chunks, the latter how the _prev_in_vs/_next_in_vs chains headers for adjacent chunks together. The latter is a result of some confusion Richard and I had during our last review round.

coleenp reviewed (approved):
I still approve this change.

coleenp commented:
@reinrich approve this change so @tstuefe can check it in!  Anything left can be addressed in a new PR.  Thanks!

reinrich reviewed (approved):
Hi Thomas,
this is a major enhancement and it looks good to me. Congratulations! :)
Cheers, Richard.

tstuefe commented:
> Hi Thomas,
> this is a major enhancement and it looks good to me. Congratulations! :)
> Cheers, Richard.

Thanks a lot Richard!

lkorinth commented:
Still looks good to me.

tstuefe commented:
/integrate
